libr/core/rtr.c

```diff
@@ -1847,68 +1848,197 @@ R_API char *r_core_rtr_cmds_query (RCore *core, const char *host, const char *po
+				break;
+			}
+		}
 	}
+	rtr_cmds_client_context *client_context = client->data;
+	uv_close ((uv_handle_t *) client, (uv_close_cb) free);
+	free (client_context->res);
+	free (client_context);
+}
 
-	s = r_socket_new (0);
-	s->local = r_config_get_i(core->config, "tcp.islocal");
+static void rtr_cmds_alloc_buffer(uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf) {
+	rtr_cmds_client_context *context = handle->data;
+	buf->base = context->buf + context->len;
+	buf->len = sizeof (context->buf) - context->len - 1;
+}
 
-	if (!r_socket_listen (s, port, NULL)) {
-		eprintf ("Error listening on port %s\n", port);
-		r_socket_free (s);
-		return false;
+static void rtr_cmds_write(uv_write_t *req, int status) {
+	rtr_cmds_client_context *context = req->data;
+
+	if (status) {
+		eprintf ("Write error: %s\n", uv_strerror (status));
 	}
 
-	eprintf ("Listening for commands on port %s\n", port);
-	listenport = port;
-	r_cons_break_push ((RConsBreak)r_core_rtr_http_stop, core);
-	for (;;) {
-		if (r_cons_is_breaked ()) {
-			break;
+	free (req);
+	rtr_cmds_client_close (context->client, true);
+}
+
+static void rtr_cmds_read(uv_stream_t *client, ssize_t nread, const uv_buf_t *buf) {
+	rtr_cmds_context *context = client->loop->data;
+	rtr_cmds_client_context *client_context = client->data;
+
+	if (nread < 0) {
+		if (nread != UV_EOF) {
+			eprintf ("Failed to read: %s\n", uv_err_name ((int) nread));
 		}
-		void *bed = r_cons_sleep_begin ();
-		ch = r_socket_accept (s);
-		buf[0] = 0;
-		ret = r_socket_read (ch, buf, sizeof (buf) - 1);
-		r_cons_sleep_end (bed);
-		if (ret > 0) {
-			buf[ret] = 0;
-			for (i = 0; buf[i]; i++) {
-				if (buf[i] == '\n') {
-					buf[i] = buf[i + 1]? ';': '\0';
-				}
-			}
-			if ((!r_config_get_i (core->config, "scr.prompt") &&
-			     !strcmp ((char *)buf, "q!")) ||
-			    !strcmp ((char *)buf, ".--")) {
-				r_socket_close (ch);
-				break;
-			}
-			str = r_core_cmd_str (core, (const char *)buf);
-			bed = r_cons_sleep_begin ();
-			if (str && *str)  {
-				r_socket_write (ch, str, strlen (str));
-			} else {
-				r_socket_write (ch, "\n", 1);
-			}
-			r_cons_sleep_end (bed);
-			free (str);
+		rtr_cmds_client_close ((uv_tcp_t *) client, true);
+		return;
+	} else if (nread == 0) {
+		return;
+	}
+
+	buf->base[nread] = '\0';
+	char *end = strchr (buf->base, '\n');
+	if (!end) {
+		return;
+	}
+	*end = '\0';
+
+	r_cons_sleep_end (context->bed);
+	client_context->res = r_core_cmd_str (client_context->core, (const char *)client_context->buf);
+	context->bed = r_cons_sleep_begin ();
+
+	if (!client_context->res || !*client_context->res) {
+		free (client_context->res);
+		client_context->res = strdup ("\n");
+	}
+
+	if (!client_context->res || (!r_config_get_i (client_context->core->config, "scr.prompt") &&
+				 !strcmp ((char *)buf, "q!")) ||
+				 !strcmp ((char *)buf, ".--")) {
+		rtr_cmds_client_close ((uv_tcp_t *) client, true);
+		return;
+	}
+
+	uv_write_t *req = R_NEW (uv_write_t);
```

Comment:
R_NEW0 ?

Reply:
Not needed, the struct is not initialized in all libuv examples I have seen.